internal/iterator/fan_in.go

```diff
@@ -10,46 +10,63 @@ import (
 )
 
 type FanIn struct {
-	ctx       context.Context
-	cancel    context.CancelFunc
-	count     int
-	out       chan *Msg
-	addCh     chan (chan *Msg)
-	mu        sync.Mutex
-	accepting bool
-	pool      *pool.ContextPool
-	drained   chan bool
+	ctx        context.Context
+	cancel     context.CancelFunc
+	out        chan *Msg
+	addCh      chan (chan *Msg)
+	drainQueue []chan *Msg
+	accepting  bool
+	mu         sync.Mutex
+	wg         sync.WaitGroup
+	pool       *pool.ContextPool
 }
 
 func NewFanIn(ctx context.Context, limit int) *FanIn {
 	ctx, cancel := context.WithCancel(ctx)
 
 	f := &FanIn{
-		ctx:       ctx,
-		cancel:    cancel,
-		count:     0,
-		out:       make(chan *Msg, limit),
-		addCh:     make(chan (chan *Msg), limit),
-		accepting: true,
-		mu:        sync.Mutex{},
-		pool:      concurrency.NewPool(ctx, limit),
-		drained:   make(chan bool, 1),
+		ctx:        ctx,
+		cancel:     cancel,
+		out:        make(chan *Msg, limit),
+		addCh:      make(chan (chan *Msg), limit),
+		drainQueue: make([]chan *Msg, 0),
+		accepting:  true,
+		mu:         sync.Mutex{},
+		wg:         sync.WaitGroup{},
+		pool:       concurrency.NewPool(ctx, limit),
 	}
 
+	f.wg.Add(1)
 	go f.run()
+
 	return f
 }
 
+func (f *FanIn) cleaner() {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	for _, ch := range f.drainQueue {
+		Drain(ch).Wait()
+	}
```

Comment:
Do we want to run all Drain(ch) first and then call Wait on them?  Right now, the second ch will not start until the first one Wait() is done.

Reply:
I don't want to end up with a bunch of concurrent goroutines being spawned just for draining channels, I'm ok with them being sequential (otherwise I wouldn't `.Wait()` on them.

Reply:
Mind adding comment to call it out specifically?